### ISSUE TITLE
IOS - Sign In - Page size is changing after use wrong 2FA and clicking on Save button

### ISSUE THREAD

**Issue by kbecciv** [Expensify/App#9817]
If you haven’t already, check out our [contributing guidelines](https://github.com/Expensify/ReactNativeChat/blob/main/docs/CONTRIBUTING.md) for onboarding and email contributors@expensify.com to request to join our Slack channel!
___

## Action Performed:
Precondition: need to use 2FA account
1. Launch the app
2. On Sign in user your username
3. Type your password
4. Type wrong 2FA
5. Tap on Sign in

## Expected Result:
Page size is not changing after use wrong 2FA and clicking on Save button

## Actual Result:
Page size is changing after use wrong 2FA and clicking on Save button

## Workaround:
Unknown

## Platform:

Where is this issue occurring?

- Web
- iOS
- Android
- Desktop App
- Mobile Web

**Version Number:** 1.1.82.5

**Reproducible in staging?:** Yes

**Reproducible in production?:** Yes

**Email or phone of affected tester (no customers):** any 2FA accounts

**Logs:** https://stackoverflow.com/c/expensify/questions/4856

**Notes/Photos/Videos:** Any additional supporting documentation

https://user-images.githubusercontent.com/93399543/178118382-13089550-7051-41f3-9950-7b689930da91.mp4

**Expensify/Expensify Issue URL:** 

**Issue reported by:** Applause - Internal Team

**Slack conversation:** 

[View all open jobs on GitHub](https://github.com/Expensify/App/issues?q=is%3Aopen+is%3Aissue+label%3A%22Help+Wanted%22)

**Comment by mvtglobally** [comment]
Issue not reproducible during KI retests. (First week)

**Comment by laurenreidexpensify** [comment]
Doublechecking if I need to add this to Upwork before it's reproducible

**Comment by laurenreidexpensify** [comment]
Holding on this to be reproduced in next week's KI testing

**Comment by laurenreidexpensify** [comment]
Holding on this to be reproduced in this week's KI testing

**Comment by laurenreidexpensify** [comment]
Looks like it was reproducible this week so moving forward

**Comment by laurenreidexpensify** [comment]
https://www.upwork.com/jobs/~014f53fc107128911d

**Comment by errahsoufiane** [comment]
### **Proposal**
https://github.com/Expensify/App/blob/151ec1ad3355cfa162ad490030f19a7134bbc076/src/pages/signin/SignInPageLayout/SignInPageContent.js#L51
This happens on devices with small screen width (isSmallScreenWidth = true) because SignInPageForm width is changing dynamically based on width sizes of child views.

**Solution:**
set SignInPageForm width statically when isSmallScreenWidth = true
Code Implementation:
https://github.com/Expensify/App/blob/151ec1ad3355cfa162ad490030f19a7134bbc076/src/styles/styles.js#L935
After this line, add this:
```
        width: 335,
```

Before fix:
![before-without-error](https://user-images.githubusercontent.com/108141252/181048872-513f4b6b-be37-4b2d-a569-c77c3318866c.png)
![before-with-error](https://user-images.githubusercontent.com/108141252/181048924-d160ac03-4e94-4413-8d12-31d7364f4cf8.png)

After fix:
![after-without-error](https://user-images.githubusercontent.com/108141252/181048998-6fafe14d-83f7-4e2e-b9cf-5cc851806719.png)
![after-with-error](https://user-images.githubusercontent.com/108141252/181049032-5f72dc44-687e-4330-a7d0-50184fe78ebe.png)

**Comment by rushatgabhane** [comment]
@errahsoufiane hmm, will all small devices use a width of 335?

**Comment by olya-ivanochko** [comment]
> @errahsoufiane hmm, will all small devices use a width of 335?

yes, I think it's no problem since that value is bigger than the smallest mobile device (android: 360, iPhone: 375)

**Comment by errahsoufiane** [comment]
> @errahsoufiane hmm, will all small devices use a width of 335?

yes I think it's no problem since it's even bigger than smallest device width (android: 360, iphone: 375)

**Comment by rushatgabhane** [comment]
@errahsoufiane cool.

In your video, I noticed that the width of the page content is still changing.

**Comment by errahsoufiane** [comment]
> @errahsoufiane cool.
> 
> In your video, I noticed that the width of the page content is still changing.

Oh, that video is not real app play but combination of 4 screenshots:
- before fix 1.png (not showing error before fix)
- before fix 2.png (showing error before fix)
- fix 1.png (not showing error after fix)
- fix 2.png (showing error after fix)

**Comment by rushatgabhane** [comment]
Oh haha, didn't notice that all.

> yes I think it's no problem since it's even bigger than smallest device width (android: 360, iphone: 375

@errahsoufiane galaxy fold has a width of 280px, so the login page would break.

**Comment by errahsoufiane** [comment]
@rushatgabhane you're right. As far as I am testing, this happens only on iOS. Android works fine. tested on multiple devices.
so my **updated proposal** is:

https://github.com/Expensify/App/blob/151ec1ad3355cfa162ad490030f19a7134bbc076/src/pages/signin/SignInPageLayout/SignInPageContent.js#L51
```
            props.isSmallScreenWidth && (Platform.OS === 'ios' ? styles.signInPageContentContainer : styles.signInPageNarrowContentContainer),
```


https://github.com/Expensify/App/blob/151ec1ad3355cfa162ad490030f19a7134bbc076/src/styles/styles.js#L935
<img width="997" alt="style" src="https://user-images.githubusercontent.com/108141252/181179904-fcc05070-c7ed-4af6-ae63-caa4de7d8adf.png">

**Comment by rushatgabhane** [comment]
@errahsoufiane i still think that setting a fixed width of 335 isn't the best idea.

iPhone 5 has a width of 320px ([source](https://yesviz.com/devices/iphone5/))

**Comment by errahsoufiane** [comment]
@rushatgabhane 
```
This happens on devices with small screen width (isSmallScreenWidth = true) because SignInPageForm width is changing dynamically based on width sizes of child views.
```
this was what I had said and this is actually true
There's no defined width of this parent view and it is set as width of a child view which has the longest width among child views.
You can see clearly what I mean from this video:
https://drive.google.com/file/d/1vVr4nv6-wdiIusbR2TboQGchQf8uCZpB/view?usp=sharing

so **updated proposal** is: similar to my first proposal - `set fixed width to 335` but only when window width is bigger than this value so that it could be responsive to very small devices like iphone 5, galaxy fold.
If width is smaller than 335, no need to set either width or maxWidth.

https://github.com/Expensify/App/blob/main/src/styles/styles.js#L935
<img width="1143" alt="styles" src="https://user-images.githubusercontent.com/108141252/181483879-fb97de9e-51cc-4417-9a04-4a2da7270580.png">

https://github.com/Expensify/App/blob/main/src/pages/signin/SignInPageLayout/SignInPageContent.js#L51
<img width="1439" alt="SignInPageContent" src="https://user-images.githubusercontent.com/108141252/181483950-29d0f04f-4445-4626-b760-c503da49a2f9.png">

**Comment by aimane-chnaif** [comment]
### **Proposal**
This happens when try to set maxWidth on alignSelf centered view without specifying width.
Here is the fix: https://stackoverflow.com/questions/39536665/react-native-alignself-center-and-stretch-to-maxwidth, this means setting `flex: 1` fixes the issue. But this can't be applied to ScrollView content directly since its direction is vertical.
So the solution is to wrap this content with horizontal view (`flexDirection: 'row'`) so that `flex: 1` can be applied inside that view.
Full code implementation in [SignInPageContent.js](https://github.com/Expensify/App/blob/main/src/pages/signin/SignInPageLayout/SignInPageContent.js):
```
const SignInPageContent = props => (
    <ScrollView
        keyboardShouldPersistTaps="handled"
        showsVerticalScrollIndicator={false}
        style={[
            styles.h100,
            !props.isSmallScreenWidth && styles.alignSelfCenter,
            !props.isSmallScreenWidth && styles.signInPageWideLeftContainer,
        ]}
        contentContainerStyle={[
            scrollViewContentContainerStyles,
            styles.alignItemsCenter,
            !props.isSmallScreenWidth && styles.ph6,
        ]}
    >
        <View style={[styles.flex1, styles.flexRow]}>
            <View style={[
                styles.flex1,
                props.isSmallScreenWidth && styles.signInPageNarrowContentContainer,
            ]}
            >
                <SignInPageForm style={[
                    styles.flex1,
                    styles.alignSelfStretch,
                    props.isSmallScreenWidth ? styles.ph5 : styles.ph4,
                ]}
                >
                    <LoginKeyboardAvoidingView
                        behavior="position"
                        contentContainerStyle={[
                            props.isSmallScreenWidth ? styles.signInPageNarrowContentMargin : {},
                            !props.isMediumScreenWidth ? styles.signInPageWideLeftContentMargin : {},
                            styles.mb3,
                            StyleUtils.getModalPaddingStyles({
                                shouldAddBottomSafeAreaPadding: true,
                                modalContainerStylePaddingBottom: 20,
                                safeAreaPaddingBottom: props.insets.bottom,
                            }),
                        ]}
                    >
                        <View style={[
                            styles.componentHeightLarge,
                            ...(props.isSmallScreenWidth ? [styles.mb2] : [styles.mt6, styles.mb5]),
                        ]}
                        >
                            <ExpensifyCashLogo
                                width={variables.componentSizeLarge}
                                height={variables.componentSizeLarge}
                            />
                        </View>
                        {props.shouldShowWelcomeText && (
                            <Text style={[styles.mv5, styles.textLabel, styles.h3]}>
                                {props.welcomeText}
                            </Text>
                        )}
                        {props.children}
                    </LoginKeyboardAvoidingView>
                </SignInPageForm>
                <View style={[styles.mb5, styles.alignSelfCenter, styles.ph5]}>
                    <TermsAndLicenses />
                </View>
            </View>
        </View>
    </ScrollView>
);
```

Here, this is the key to fix the main problem of this GH issue:
<img width="805" alt="flex1" src="https://user-images.githubusercontent.com/96077027/181503269-4f9b6951-a65a-4378-8da4-d23eb6ddded8.png">

**Comment by rushatgabhane** [comment]
@aimane-chnaif awesome! You're right about the root cause and I like your proposal.

@marcochavezf C+ reviewed 🎀 👀 🎀

**Comment by aimane-chnaif** [comment]
<img width="280" alt="280x653" src="https://user-images.githubusercontent.com/96077027/182111837-12c45a62-0e3d-43f4-8391-93fef71b326a.png">

**Comment by rushatgabhane** [comment]
@aimane-chnaif Thanks for the screenshot for 280px 

@marcochavezf I like @aimane-chnaif's [proposal](https://github.com/Expensify/App/issues/9817#issuecomment-1198069662)!

🎀 👀 🎀 C+ reviewed

**Comment by marcochavezf** [comment]
Cool, thanks guys! Assigning @aimane-chnaif!

**Comment by laurenreidexpensify** [comment]
@jliexpensify 👋 am OOO til 12 August so reassigning. Thanks!

**Comment by jliexpensify** [comment]
Listing: https://www.upwork.com/ab/applicants/1551917570399793152/job-details

Invited @rushatgabhane and @aimane-chnaif - cheers!

**Comment by jliexpensify** [comment]
I've had to take some unexpected time off, so assigning to Kadie (my leave buddy) to manage!

**Comment by aimane-chnaif** [comment]
Hi @kadiealexander, how's it going? What's the next step for me?

**Comment by kadiealexander** [comment]
G'day @aimane-chnaif! Next steps will be to:

 - Accept the Upwork contract I just sent across
 - Get started on submitting a PR based on your proposal
 
You can check out our [Contributing.MD](https://github.com/Expensify/App/blob/main/contributingGuides/CONTRIBUTING.md) page for more detailed instructions on working on Expensify jobs. Let me know if you have any questions!

**Comment by kadiealexander** [comment]
Not overdue.

**Comment by kadiealexander** [comment]
Everyone is paid! Thanks team :)
